pkg/systemd/services/services.jsx

```diff
@@ -70,16 +71,43 @@
 }
 
 export function updateTime() {
-    cockpit.spawn(["cat", "/proc/uptime"])
-            .then(function(contents) {
-                // first number is time since boot in seconds with two fractional digits
-                const uptime = parseFloat(contents.split(' ')[0]);
-                clock_monotonic_now = parseInt(uptime * 1000000, 10);
-            }, ex => console.log(ex.toString()));
-    cockpit.spawn(["date", "+%s"])
-            .then(time => {
-                clock_realtime_now = parseInt(time, 10) * 1000;
-            }, ex => console.log(ex.toString()));
+    // To correctly interpret monotonic timers, we need to read CLOCK_MONOTONIC. This cannot be read with shell tools
+    python.spawn("import time; print(time.clock_gettime_ns(time.CLOCK_REALTIME), time.clock_gettime_ns(time.CLOCK_MONOTONIC))",
+                 null, { err: "message" })
+            .then(output => {
+                const [realtime, monotonic] = output.split(' ').map(ns => parseInt(ns));
+                clock_realtime_now = realtime / 1000000;
+                monotonic_timer_base = (realtime - monotonic) / 1000;
+                debug("Read clocks with Python; realtime", clock_realtime_now, "monotonic base", monotonic_timer_base);
+            })
+            .catch(ex => {
+                /* If Python is not available, fall back to reading CLOCK_BOOTTIME from /proc/timer-list.
+                 * This is readable by root only */
+                console.log("Failed to read clocks with Python, using fallback:", ex.toString());
+                Promise.allSettled([
+                    cockpit.spawn(["date", "+%s"]),
+                    cockpit.file("/proc/timer_list", { superuser: "try" }).read(),
+                ])
+                        .then(([date_result, timer_list_result]) => {
+                            if (date_result.status === "fulfilled") {
+                                clock_realtime_now = parseInt(date_result.value) * 1000;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18466-20230309-111146-2a80128f-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)